File: Offer2/0037/main.go
package main

import (
	"strconv"
	"strings"
)

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func serialize(root *TreeNode) string {
	if root == nil {
		return "[]"
	}
	queue := []*TreeNode{root}
	res := []string{}
	var node *TreeNode
	for len(queue) != 0 {
		node = queue[0]
		queue = queue[1:]
		if node == nil {
			res = append(res, "null")
		} else {
			res = append(res, strconv.Itoa(node.Val))
			queue = append(queue, node.Left, node.Right)
		}
	}
	return "[" + strings.Join(res, ",") + "]"
}

func deserialize(data string) *TreeNode {
	if data == "[]" {
		return nil
	}
	res := strings.Split(data[1:len(data)-1], ",")
	val, _ := strconv.Atoi(res[0])
	root := &TreeNode{val, nil, nil}
	queue := []*TreeNode{root}
	i := 1
	var node *TreeNode
	for len(queue) != 0 {
		node = queue[0]
		queue = queue[1:]
		if res[i] != "null" {
			val, _ = strconv.Atoi(res[i])
			node.Left = &TreeNode{val, nil, nil}
			queue = append(queue, node.Left)
		}
		i += 1
		if res[i] != "null" {
			val, _ = strconv.Atoi(res[i])
			node.Right = &TreeNode{val, nil, nil}
			queue = append(queue, node.Right)
		}
		i += 1
	}
	return root
}

File: Offer2/0037/main.py
import collections
from typing import List


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Codec:

    def serialize(self, root: TreeNode) -> str:
        if root == None: return "[]"
        queue: collections.deque[TreeNode] = collections.deque()
        queue.append(root)
        res: List[str] = []
        while len(queue) != 0:
            node = queue.popleft()
            if node == "None":
                res.append("null")
            else:
                res.append(str(node.val))
                queue.append(node.left)
                queue.append(node.right)
        return "[" + ",".join(res) + "]"

    def deserialize(self, data: str) -> TreeNode:
        if data == "[]": return None
        res: List[str] = data[1:-1].split(",")
        root: TreeNode = TreeNode(int(res[0]))
        queue: collections.deque[TreeNode] = collections.deque()
        queue.append(root)
        i = 1
        while len(queue) != 0:
            node = queue.popleft()
            if res[i] != "null":
                node.left = TreeNode(int(res[i]))
                queue.append(node.left)
            i += 1
            if res[i] != "null":
                node.right = TreeNode(int(res[i]))
                queue.append(node.right)
            i += 1
        return root